driver/src/deposit_driver.rs

```diff
@@ -21,74 +23,41 @@ pub fn run_deposit_listener<D, C>(db: &D, contract: &C) -> Result<(bool), Driver
-
-    println!(
-        "Current block is {:?} and the last deposit_ind_creationBlock is {:?}",
-        current_block, current_deposit_ind_block
-    );
-
-    // if 20 blocks have past since the first deposit and we are not in the newest slot, we apply the deposit.
-    if current_deposit_ind_block + 20 < current_block
-        && deposit_ind != current_deposit_ind + 1
-    {
-        println!("Next deposit_slot to be processed is {}", deposit_ind);
-        let deposits = db.get_deposits_of_slot(deposit_ind.low_u32())?;
-        println!("Amount of deposits to be processed{:?}", deposits.len());
-        //rehash deposits
-        let deposit_hash: H256 = deposits.rolling_hash();
-
-        if deposit_hash != deposit_hash_pulled {
-            return Err(DriverError::new(
-                &format!("Pending deposit hash from contract ({}), didn't match the one found in db ({})", 
-                deposit_hash, deposit_hash_pulled), ErrorKind::StateError
-            ));
-        }
-
-        if deposit_slot_empty && deposit_ind != U256::zero() {
-            println!("deposit_slot {} already processed", deposit_ind);
-        } else {
-            // calculate new state by applying all deposits
-            state = apply_deposits(&mut state, &deposits);
-            println!("New State_hash is{:?}", state.rolling_hash());
-
-            //send new state into blockchain
-            //applyDeposits signature is (slot, _currStateRoot, _newStateRoot, deposit_slotHash)
-            let slot = U256::from(deposit_ind);
-            let _curr_state_root = curr_state_root;
-            let _new_state_root = H256::from(state.rolling_hash());
-
-            contract.apply_deposits(slot, _curr_state_root, _new_state_root, deposit_hash_pulled)?;
+    let deposit_slot = contract.get_current_deposit_slot()?;
+
+    println!("Current top deposit_slot is {:?}", deposit_slot);
+    let slot = util::find_first_unapplied_slot(
+        deposit_slot + 1, 
+        Box::new(&|i| contract.has_deposit_slot_been_applied(i))
+    )?;
+    if slot <= deposit_slot {
+        println!("Highest unprocessed deposit_slot is {:?}", slot);
+        if util::can_process(slot, contract,
+            Box::new(&|i| contract.creation_block_for_deposit_slot(i))
+        )? {
+            println!("Processing deposit_slot {:?}", slot);
+            let state_root = contract.get_current_state_root()?;
+            let contract_deposit_hash = contract.deposit_hash_for_slot(slot)?;
+            let mut balances = db.get_current_balances(&state_root)?;
+
+            let deposits = db.get_deposits_of_slot(slot.low_u32())?;
+            let deposit_hash = deposits.rolling_hash();
+            if deposit_hash != contract_deposit_hash {
+                return Err(DriverError::new(
+                    &format!("Pending deposit hash from contract ({}), didn't match the one found in db ({})", 
+                    deposit_hash, contract_deposit_hash), ErrorKind::StateError
```

Comment:
The order here is off. According to the string we have to print `contract_deposit_hash` first